File: backend/src/domain/repositories/user.repository.ts
import { User } from '../entities/user.entity'

export interface UserRepository {
  findAll(): Promise<User[]>
  findById(id: number): Promise<User | null>
  findByEmail(email: string): Promise<User | null>
  create(user: User): Promise<User>
  update(user: User): Promise<User>
  remove(id: number): Promise<User | null>
  nextId(): Promise<number>
}

export const USER_REPOSITORY = 'USER_REPOSITORY'

File: go-microservice/internal/domain/repositories/user_repository.go
package repositories

import (
	"context"

	"github.com/google/uuid"
	"github.com/southern-martin/go-microservice/internal/domain/entities"
)

type UserRepository interface {
	Create(ctx context.Context, user *entities.User) error
	GetByID(ctx context.Context, id uuid.UUID) (*entities.User, error)
	GetByEmail(ctx context.Context, email string) (*entities.User, error)
	Update(ctx context.Context, user *entities.User) error
	Delete(ctx context.Context, id uuid.UUID) error
	List(ctx context.Context, limit, offset int, search string) ([]*entities.User, error)
	Count(ctx context.Context, search string) (int64, error)
}
